lib/puppet/parser/functions/delete.rb

```diff
@@ -20,19 +20,28 @@ module Puppet::Parser::Functions
 
     delete('abracadabra', 'bra')
     Would return: 'acada'
+
+    delete(['abracadabra'], '^.*bra.*$', true)
+    Would return: []
+
+    delete(['abracadabra'], '^.*jimbob.*$', true)
+    Would return: ['abracadabra']
   EOS
   ) do |arguments|
 
-    if (arguments.size != 2) then
-      raise(Puppet::ParseError, "delete(): Wrong number of arguments "+
-                                  "given #{arguments.size} for 2.")
-    end
+    raise(Puppet::ParseError, "delete(): Wrong number of arguments "+
+                              "given #{arguments.size} for 2 or 3.") unless (2..3).include? arguments.size
 
     collection = arguments[0].dup
+    use_regex = arguments.size == 3 ? arguments[2] : false
     Array(arguments[1]).each do |item|
       case collection
         when Array, Hash
-          collection.delete_if { |coll_item| coll_item =~ %r{#{item}} }
+          if use_regex
+            collection.reject! { |coll_item| (coll_item =~ %r{#{item}}) } # make no assumptions about regex format
+          else
+            collection.reject! { |coll_item| coll_item == item }
```

Comment:
I don't believe it did, but I'm unable to verify at the moment. If you want the fuzzy match functionality I can wrap it in ^.*escapedvar.*$
 
Is that amenable?

> On Apr 21, 2016, at 5:36 PM, Bryan Jen <notifications@github.com> wrote:
> 
> In lib/puppet/parser/functions/delete.rb:
> 
> >      Array(arguments[1]).each do |item|
> >        case collection
> >          when Array, Hash
> > -          collection.delete_if { |coll_item| coll_item =~ %r{#{item}} }
> > +          if use_regex
> > +            collection.reject! { |coll_item| (coll_item =~ %r{#{item}}) } # make no assumptions about regex format
> > +          else
> > +            collection.reject! { |coll_item| coll_item == item }
> it's the same idea for Arrays and Hashes, I believe. The old behavior will delete the key/value pair based on a fuzzy match of the substring. The updated code now requires an absolute match.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>